[SPARK-6994] Allow to fetch field values by name in sql.Row

vidma opened the pull request:
It looked weird that up to now there was no way in Spark's Scala API to access fields of `DataFrame/sql.Row` by name, only by their index.

This tries to solve this issue.

Commit:
[SPARK-6994][SQL] Allow to fetch field values by name on Row

- add fieldIndex(name: String)
- add getAs[T](fieldName: String)
- add getValuesMap[T] returning a map of values for the requested fieldNames

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala:
```diff
@@ -0,0 +1,43 @@
+
+object RowImplicits {
+
+  implicit class RowWithNamedFields[K, V](sqlRow: Row) {
```
why don't we just add this to Row itself?

vidma replied:
:+1: if so, I guess I should rename `getValue[T]` into an overloaded `getAs[T](fieldName: String)` to be consistent with an existing method `getAs`? (I remember overloading didn't work with implicits...)

rxin replied:
yup sounds good

vidma replied:
For the new method to be convenient to use it should go into `Row`. 
However, the schema is not always defined for a `Row`, so `Row.getAs[T](fieldName: String)` with throw an `UnsupportedOperationException`.
`GenericRowWithSchema` will implement the case when schema exists.

Commit:
[SPARK-6994][SQL] Add fieldIndex to schema (StructType)

vidma commented:
moved the changes to the `Row` and `RowWithSchema` classes, and also added a `fieldIndex` to the schema (`StructType`).

also got rid of dependency on HiveTest.

vidma commented:
it's slightly weird that there exists a `RowSuite.scala` in `sql-core` and none in `sql-catalyst`.
Anyway, the one in `sql-core` looks like integration test, so I added a little test there too, to make sure it works on Rows created via DataFrames.

marmbrus commented:
ok to test

SparkQA commented:
[Test build #30611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30611/consoleFull) for   PR 5573 at commit [`6145ae3`](https://github.com/apache/spark/commit/6145ae3fb30251e0affdda2358f2c95d1407d3aa).

SparkQA commented:
[Test build #30611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30611/consoleFull) for   PR 5573 at commit [`6145ae3`](https://github.com/apache/spark/commit/6145ae3fb30251e0affdda2358f2c95d1407d3aa).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30611/
Test PASSed.

marmbrus commented:
Thanks, merged to master.
